### ISSUE TITLE
v0.37.0.2 doesn't sync Vertica schema

### ISSUE THREAD

**Issue by grkhr** [metabase/metabase#13628]
**Describe the bug**
When I sync Vertica in admin panel (`Sync database schema now` button), new tables/fields don't appear in Data Model. But from SQL I can access it. On Metabase **v0.35.3** it was fine.

Also I've tried to restart Metabase, restart Vertica and reconnect Vertica, now schema is **empty** in Data Model and still accessible from SQL. Logs seem to be fine. 

We are using `vertica-jdbc-9.1.1-0.jar` and Vertica **9.1.1**.

**Logs**
```
[e92e7e38-ceb9-42ec-a521-b24a4540fe11] 2020-10-29T20:26:43+03:00 INFO metabase.sync.util STARTING: Analyze data for vertica Database 77 'vertica_test'
[e92e7e38-ceb9-42ec-a521-b24a4540fe11] 2020-10-29T20:26:43+03:00 INFO metabase.sync.util FINISHED: step 'classify-fields' for vertica Database 77 'vertica_test' (95.0 µs)
[e92e7e38-ceb9-42ec-a521-b24a4540fe11] 2020-10-29T20:26:43+03:00 INFO metabase.sync.util FINISHED: step 'fingerprint-fields' for vertica Database 77 'vertica_test' (191.7 µs)
[e92e7e38-ceb9-42ec-a521-b24a4540fe11] 2020-10-29T20:26:43+03:00 INFO metabase.sync.util STARTING: step 'classify-fields' for vertica Database 77 'vertica_test'
[e92e7e38-ceb9-42ec-a521-b24a4540fe11] 2020-10-29T20:26:43+03:00 INFO metabase.sync.util STARTING: step 'fingerprint-fields' for vertica Database 77 'vertica_test'
[e92e7e38-ceb9-42ec-a521-b24a4540fe11] 2020-10-29T20:26:43+03:00 INFO metabase.sync.util FINISHED: step 'classify-tables' for vertica Database 77 'vertica_test' (108.6 µs)
[e92e7e38-ceb9-42ec-a521-b24a4540fe11] 2020-10-29T20:26:43+03:00 INFO metabase.sync.util STARTING: step 'classify-tables' for vertica Database 77 'vertica_test'
[e92e7e38-ceb9-42ec-a521-b24a4540fe11] 2020-10-29T20:26:44+03:00 INFO metabase.sync.util FINISHED: Analyze data for vertica Database 77 'vertica_test' (451.2 ms)
```

```
{"total-tables":0,"tables-classified":0}
```

**To Reproduce**
1. Run 0.37.0.2 Metabase
2. Connect Vertica 9.1.1
3. Click "Sync database schema now"
4. Click "Re-scan field values now" (if you want)
5. Wait a minute
6. Go to "Data Model" and see empty schema

**Expected behavior**
New tables and fields will display.

**Screenshots**
![image](https://user-images.githubusercontent.com/40689054/97611208-e94ab980-1a26-11eb-8644-2ce266f43437.png)
![image](https://user-images.githubusercontent.com/40689054/97611088-c4eedd00-1a26-11eb-9092-2ea7e367516e.png)

**Severity**
Extremely severe. Can't use custom question and make dashboards from new tables.

**Metabase Diagnostic Info**

```json
{
  "browser-info": {
    "language": "en",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.80 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.7+10",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.7",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.7+10",
    "os.name": "Linux",
    "os.version": "4.4.0-137-generic",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "bigquery",
      "mysql",
      "vertica",
      "postgres",
      "presto"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "11.2 (Debian 11.2-1.pgdg90+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-10-26",
      "tag": "v0.37.0.2",
      "branch": "release-x.37.x",
      "hash": "ba7be09"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Comment by grkhr** [comment]
Also "public" schema missing in original Vertica connection.

![image](https://user-images.githubusercontent.com/40689054/97612322-57dc4700-1a28-11eb-9cbd-f4c1f39c3151.png)

![image](https://user-images.githubusercontent.com/40689054/97612364-64f93600-1a28-11eb-9402-7d5915e189c4.png)

**Comment by dpsutton** [comment]
Hi I was unable to reproduce. I'm using `docker pull sumitchawla/vertica`

```bash
docker pull sumitchawla/vertica
docker run --rm -d -p 5433:5433 --name vertica sumitchawla/vertica
docker exec -it vertica bash
```

I used the admin tool (`admintools`) to stop the default `docker` database and then created the sample database with bout 5 million rows:

```bash
dbadmin@ba21382e27c7:/opt/vertica/examples$ /opt/vertica/sbin/install_example VMart
Installing VMart example database
Thu Oct 29 19:04:04 UTC 2020
Creating Database
Completed
Generating Data. This may take a few minutes.
Completed
Creating schema
Completed
Loading 5 million rows of data. Please stand by.
Completed
Removing generated data files
Example database creation complete

Log is located in /opt/vertica/examples/log/ExampleInstall.txt

Thu Oct 29 19:05:19 UTC 2020
```

Once done, Metabase seems to sync fine for me:
<img width="1498" alt="image" src="https://user-images.githubusercontent.com/6377293/97624302-e636d580-19f4-11eb-8b15-7bdc954c2b88.png">

I'm using `vertica-jdbc-10.0.1-0.jar` in my plugins folder downloaded from vertica's download page.

If you can spot anything that might be different that would be helpful. I'm gonna remove this, sync it in 36.0 and then run 37 and see if it gets lost in the upgrade but wondering what might be different.

**Comment by grkhr** [comment]
Hi! Thanks for your reply.

As I said, we are using Vertica **9.1.1** with `vertica-jdbc-9.1.1-0.jar` and were upgrading from Metabase **v0.35.3**.

In [docs](https://www.metabase.com/docs/latest/administration-guide/databases/vertica.html) on Metabase's official site says that it should works fine from 8.0.0 version of Vertica.

**Comment by dacort** [comment]
@grkhr It's possible this is a permissions issue - can you let us know what user is being used to connect to Vertica and what their permissions are? And if you're able to connect with the `dbadmin` user does it work?

**Comment by grkhr** [comment]
It was connected via dbadmin all this time :(

![image](https://user-images.githubusercontent.com/40689054/97634413-b3b5c880-1a46-11eb-86aa-4e188371bda8.png)

**Comment by dpsutton** [comment]
hi @grkhr 

I've been trying to recreate today.

Best i can come up:
on v0.36.3, I add the sample vertica db. I go to explore the data and its not there. I go to admin and hit the sync button and it syncs. If i go back to explore the data, the schema still appears empty. But if i refresh it is there. I was never able to recreate this on v0.37.0.2 unfortunately.

**Comment by dacort** [comment]
I've been poking a little bit with this as well, but unable to recreate. I also tried with Vertica 9.2 (couldn't find a 9.1 Docker image).

@grkhr We may need some more debug details from the sync process. It's a little tricky because 0.37 uses Log4jv2 and 0.35 uses Log4jv1, but here's what you can do.

1. As @dpsutton mentioned, make sure you fresh the data model page. Changes don't automatically show up there in some cases.
2. In 0.37.x, run Metabase with additional debug logging for `metabase.sync` namespace: `java -Dlog4j.configurationFile="https://log4j.us/v2/templates/metabase?trace=metabase.sync" -jar metabase.jar` (or Docker `-e JAVA_OPTS="-Dlog4j.configurationFile=https://log4j.us/v2/templates/metabase?trace=metabase.sync"`)
3. If you want to get additional debug details in 0.35, you'll need to change the log4j config a little bit - note `configuration` and the different URL: `java -Dlog4j.configuration="https://log4j.us/templates/metabase?trace=metabase.sync" -jar metabase.jar`

**Comment by dpsutton** [comment]
Also, we have logging in the UI. Admin > Troubleshooting > Tasks
<img width="1521" alt="image" src="https://user-images.githubusercontent.com/6377293/97642811-1a6dbe80-1a14-11eb-8ea8-36744c1e2d20.png">

---
---

<img width="1130" alt="image" src="https://user-images.githubusercontent.com/6377293/97642835-25c0ea00-1a14-11eb-98f5-8df6890189d1.png">

---
---
<img width="646" alt="image" src="https://user-images.githubusercontent.com/6377293/97642791-10e45680-1a14-11eb-96e6-72b037e38c92.png">

**Comment by grkhr** [comment]
Well, we've figured it out.

The problem was in driver. Replacing `vertica-jdbc-9.1.1-0.jar` to `vertica-jdbc-10.0.1-0.jar` fixed the problem for us.

Here is steps if you still want to reproduce the bug:

1. Run Verica 9.1.1 with this dockerfile:
```
FROM ubuntu:14.04
MAINTAINER Jean Baptiste Favre <docker@jbfavre.org>

ARG VERTICA_PACKAGE="vertica_9.1.1-0_amd64.deb"

ENV SHELL "/bin/bash"
ENV DEBIAN_FRONTEND noninteractive
ENV TERM 1

ADD ${VERTICA_PACKAGE} /tmp/
#ADD scripts/debian_cleaner.sh /tmp/

RUN /usr/bin/apt-get update -yqq \
 && /usr/bin/apt-get upgrade --no-install-recommends -yqq \
 && /usr/bin/apt-get install --no-install-recommends -yqq curl ca-certificates \
 && /usr/bin/chsh -s /bin/bash root \
 && /bin/rm /bin/sh && ln -s /bin/bash /bin/sh \
 && /usr/bin/curl -o /usr/local/bin/gosu -SL 'https://github.com/tianon/gosu/releases/download/1.1/gosu' \
 && /bin/chmod +x /usr/local/bin/gosu \
 && /usr/sbin/groupadd -r verticadba \
 && /usr/sbin/useradd -r -m -s /bin/bash -g verticadba dbadmin \
 && /usr/local/bin/gosu dbadmin mkdir /tmp/.python-eggs \
 && /usr/sbin/locale-gen en_US en_US.UTF-8 \
 && /usr/sbin/dpkg-reconfigure locales \
 && /usr/bin/apt-get install --no-install-recommends -yqq openssh-server openssh-client mcelog sysstat dialog libexpat1 \
 && /usr/bin/dpkg -i /tmp/${VERTICA_PACKAGE} \
 && /opt/vertica/sbin/install_vertica --license CE --accept-eula --hosts 127.0.0.1 \
                                      --dba-user-password-disabled --failure-threshold NONE --no-system-configuration \
 && /usr/bin/apt-get remove --purge -y curl ca-certificates libpython2.7
# && /bin/bash /tmp/debian_cleaner.sh

ENV PYTHON_EGG_CACHE /tmp/.python-eggs
ENV VERTICADATA /home/dbadmin/docker
VOLUME /home/dbadmin/docker
ENTRYPOINT ["/opt/vertica/bin/docker-entrypoint.sh"]
ADD ./docker-entrypoint.sh /opt/vertica/bin/
RUN chmod +x /opt/vertica/bin/docker-entrypoint.sh

EXPOSE 5433
```
2. Download [9.1.1 driver](https://www.vertica.com/client_drivers/9.1.x/9.1.1-0/vertica-jdbc-9.1.1-0.jar)
3. Run Metabase v0.35.3
4. Upgrade to v0.37.0.2
5. Reconnect Vertica
6. Try to sync schema

**Comment by grkhr** [comment]
Sorry for the misinformation, a problem wasn't with the driver. After upgrading to `vertica-jdbc-10.0.1-0.jar` the problem was still happening the next day.

The real problem in new schema-sync-flow. Our Vertica is large (300gb) and because of some heavy views and tables it was long: about **2 hours on v0.37.0.2** against **1-2 minutes on v0.35.3**. 

Main use case: when I update table or view columns, I need to use a new table schema instantly in custom questions / use easy-filters.

**How I found it**
I've created a test user with permission on a small schema and ran syncing on v0.35.3 and v0.37.0.2. 
Here is full query-logs:
[v0-35-3-query-log.csv.zip](https://github.com/metabase/metabase/files/5478539/v0-35-3-query-log.csv.zip)
[v0-37-0-2-query-log.csv.zip](https://github.com/metabase/metabase/files/5478545/v0-37-0-2-query-log.csv.zip)


**v0.35.3** getting table's schemas from `v_catalog.odbc_columns`, it works fast and columns appear in Data Model instantly after this step. And after Metabase is doing queries like `select {columns} from {table} limit 10000`.

**v0.37.0.2** for some reason doing queries like `select * from {table} limit 1`, **then** getting columns from `v_catalog.odbc_columns` and only after this columns appear in Data Model. It works well on small databases, but not if you have large tables and a lot of views.

Seems that it's happening because of this commit https://github.com/metabase/metabase/commit/0c83c28e266706776cef979d36b78831f6d49d44#diff-fcbc67149f48554b35e85208009bc7cd27be9a752357f774eee1e652cbe0f9ba

**Comment by dacort** [comment]
@grkhr Ahh, ok. Thank you so much for the additional investigative work here.

The `select * from {table} limit 1` queries were added as part of [this PR](https://github.com/metabase/metabase/pull/10892) that was intended to have Metabase only display schemas/tables for which the user _actually_ has access to. Clearly this is sub-optimal for larger datasets.

This has shown up in a couple other issues (#13650 and maybe #13611), so we're planning to address it in an upcoming patch release.

**Comment by dacort** [comment]
@grkhr Can you provide a couple more details here?

1. How many tables/views do you have in your database?
2. Do the `LIMIT 1` queries take a long time to execute? e.g. are they doing a partition scan similar to #13650?

**Comment by grkhr** [comment]
@dacort 

1.  **Tables**: 180
**Views**: 150
**Vertica storage**: around 300gb
**Disk info**: TOSHIBA MG04ACA400EY, SATA Version is:  SATA 3.0, 6.0 Gb/s (current: 6.0 Gb/s)
**Server info**: Intel(R) Xeon(R) CPU E5-1650 v3 @ 3.50GHz, 12 cores, 256GB RAM

2. Yes, `LIMIT 1` queries take a long time on views with heavy SQL definition queries. But with `WHERE 1 <> 1` clause it works fast. Just tested it on heavy view, ~**80ms** against **37s**.

**Comment by camsaul** [comment]
Fixed by #13746
